### PR TITLE
Issue_1957 Returning the Matplotlib plt object in the plot_bounding_box_gallery.py file

### PR DESCRIPTION
**Pull request by Zekrom-7780**
# What does this PR do?

Fixes Part B of Issue #1957 :- Returning the Matplotlib plt object in the `plot_bounding_box_gallery.py` file

This was discussed in the Issue   #1957

I have added `import matplotlib.pyplot as plt` in the `plot_bounding_box_gallery.py` file

I have written no new tests for this, since when I ran `pytest plot_bounding_box_gallery.py`, there were no tests

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 Also tagged
@ianstenbit  @jbischof @innat

### CODE REVIEW

**Comment by innat** [comment]
@Zekrom-7780 
You may need to change these types of [logic](https://github.com/Zekrom-7780/keras-cv/blob/b5a868e62c5bd124fdf1f3f6233f015573e64d86/keras_cv/visualization/plot_image_gallery.py#L120-L127).

As mentioned by ian, [here](https://github.com/keras-team/keras-cv/issues/1957#issuecomment-1640484693), if saving path and show, both are none, only then return the figure object.

```python
filepath = tf.keras.utils.get_file(origin="https://i.imgur.com/9i63gLN.jpg")
image = keras.utils.load_img(filepath)
image = np.array(image)[None,...]
plt_obj = keras_cv.visualization.plot_image_gallery(
    image, 
    rows=1, 
    cols=1, 
    value_range=(0, 255), 
    show=False, 
    path=None, 
    scale=4
)
```

In torch, we can do with [`torchvision.utils.make_grid`](https://pytorch.org/vision/main/generated/torchvision.utils.make_grid.html)

```python
fig, ax = plt.subplots(figsize=(12, 6))
plt.title("Sample images from CIFAR10 dataset")
ax.set_xticks([])
ax.set_yticks([])
ax.imshow(make_grid(images, nrow=16).permute(1, 2, 0))
plt.show()
```

**Comment by Zekrom-7780** [comment]
Ok @innat, 
I understood that ONLY WHEN `if path is None and show is None` I need to return the matplotlib fig, and in the rest of the cases, Should I leave the code unchanged and not return the fig?

File: keras_cv/visualization/plot_image_gallery.py [hunk 1]
```diff
@@ -179,16 +179,18 @@ def plot_image_gallery(
             current_axis.axis("off")
 
     if path is None and not show:
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:181]: I think this behavior was correct before -- we only return the plot if the user isn't showing or saving it

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by innat** [comment]
> Ok @innat, I understood that ONLY WHEN `if path is None and show is None` I need to return the matplotlib fig, and in the rest of the cases, Should I leave the code unchanged and not return the fig?

I made a few changes to [this method](https://github.com/innat/keras-cv/blob/master/keras_cv/visualization/plot_image_gallery.py) to demonstrate to you. Take a look. I didn't consider cases, but please consider (`show=True`, path='2.png'`, or anything.).

```python
    """
    assert_matplotlib_installed("plot_bounding_box_gallery")

    if path is None and not show:
        plt.ioff()

    ....

    if path is None and not show:
        return fig, axes
```

Expected output

```python
image = keras.utils.load_img(filepath)
image = np.array(image)
image = np.repeat(image[None, ...], 6, axis=0)

catch_fig, catch_axis = visualization.plot_image_gallery(
    image, rows=2, cols=3, value_range=(0, 255),
    scale=4, show=None, path=None
)

catch_fig.suptitle('Figure Title')
catch_axis[0, 0].set_title('1 Plot')
catch_axis[0, 1].set_title('2 Plot')
catch_axis[0, 2].set_title('3 Plot')
catch_axis[1, 0].set_title('4 Plot')
catch_axis[1, 1].set_title('5 Plot')
catch_axis[1, 2].set_title('6 Plot')
catch_fig
```

download

**Comment by Zekrom-7780** [comment]
@ianstenbit  @jbischof  @innat Please check this, 
All the things that you wanted are done, 
And @innat your example that you shared is working on my code as well, and I did test with setting 
`path 'examples/visualization/'` and trying with all possibe Show choices, `Show=None`, `show-True` and `Show=False`

File: keras_cv/visualization/plot_image_gallery.py [hunk 2]
```diff
@@ -117,9 +117,9 @@ def plot_image_gallery(
     """
     assert_matplotlib_installed("plot_bounding_box_gallery")
 
-    if path is None and show is None:
-        # Default to showing the image
-        show = True
+    if path is None and not show:
+        plt.ioff()
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:121]: Why do we want to make this non-interactive?

File: keras_cv/visualization/plot_image_gallery.py [hunk 3]
```diff
@@ -178,8 +178,15 @@ def plot_image_gallery(
             current_axis.margins(x=0, y=0)
             current_axis.axis("off")
 
+    if path is None and show is None:
+        #If path is None and show is None
+        # Default to showing the image
+        show = True
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:184]: Why set this if we're just going to return immediately?

File: keras_cv/visualization/plot_bounding_box_gallery.py [hunk 4]
```diff
@@ -174,11 +174,4 @@ def unpackage_tfds_inputs(inputs):
...
 
-    plot_image_gallery(
-        plotted_images,
-        value_range,
-        legend_handles=legend_handles,
-        rows=rows,
-        cols=cols,
-        **kwargs
-    )
+    return plot_image_gallery(...)
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_bounding_box_gallery.py:177]: Need to include all of the params to `plot_image_gallery` -- I don't think this works unless it's some sort of Python wizardry 🧙

File: keras_cv/visualization/plot_image_gallery.py [hunk 5]
```diff
@@ -178,8 +178,15 @@ def plot_image_gallery(
...
+        return fig,axes
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:185]: Why return the axes?

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/visualization/plot_image_gallery.py (hunk 5 above)
Comment (Zekrom-7780) [reply keras_cv/visualization/plot_image_gallery.py:185]: Sure, will return only the image

File: keras_cv/visualization/plot_bounding_box_gallery.py (hunk 4 above)
Comment (Zekrom-7780) [reply keras_cv/visualization/plot_bounding_box_gallery.py:177]: I honestly don't know, but like the examples ran using this updated code😅, will return back to the one with parameters

File: keras_cv/visualization/plot_image_gallery.py (hunk 2 above)
Comment (Zekrom-7780) [reply keras_cv/visualization/plot_image_gallery.py:121]: Sure, won't remove the interactive mode

**Comment by Zekrom-7780** [comment]
@ianstenbit @jbischof @innat The Final changes are made, 
Also SOO SORRY for making you guys literally spell out what to do for this basic issue, I will try my best to improve from my side

File: keras_cv/visualization/plot_image_gallery.py [hunk 6]
```diff
@@ -178,8 +176,14 @@ def plot_image_gallery(
             current_axis.margins(x=0, y=0)
             current_axis.axis("off")
 
+    if path is None and show is None:
+        #If path is None and show is None
+        # Default to showing the image
+        return fig
+    
     if path is None and not show:
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:184]: We can probably drop this conditional now since the one above will cover the same case

**Review by ianstenbit — APPROVED** [review]
Looks good after one small change -- thank you!

**Comment by ianstenbit** [comment]
> @ianstenbit @jbischof @innat The Final changes are made, Also SOO SORRY for making you guys literally spell out what to do for this basic issue, I will try my best to improve from my side

No worries -- it always takes some getting used to when contributing to something new!

File: keras_cv/visualization/plot_image_gallery.py [hunk 7]
```diff
@@ -178,8 +176,14 @@ def plot_image_gallery(
             current_axis.margins(x=0, y=0)
             current_axis.axis("off")
 
+    if path is None and show is None:
+        #If path is None and show is None
```
Comment (ianstenbit) [inline keras_cv/visualization/plot_image_gallery.py:180]: Let's also remove these comments as they're not relevant / correct now

**Comment by ianstenbit** [comment]
Please take a look at the lint error -- you can run `sh shell/format.sh` and `sh shell/lint.sh` from the root directory of the repo to format and verify linting

**Comment by Zekrom-7780** [comment]
@ianstenbit So the linting part from my side is done

**Comment by Zekrom-7780** [comment]
@ianstenbit Could you please merge this pull request so that I can close this?

**Comment by ianstenbit** [comment]
@Zekrom-7780 the linter is still unhappy -- please take a look at the CI result

**Comment by Zekrom-7780** [comment]
@ianstenbit  Now the linting is done, the changes have been made

**Comment by Zekrom-7780** [comment]
@ianstenbit Linting is also happy right now it seems

File: keras_cv/visualization/plot_image_gallery.py (hunk 6 above)
Comment (ianstenbit) [reply keras_cv/visualization/plot_image_gallery.py:181]: @Zekrom-7780 it looks like this never got removed. Other than that this LGTM

**Review by ianstenbit — CHANGES_REQUESTED** [review]
See in-line comment about duplicated `if` statement

**Comment by Zekrom-7780** [comment]
@ianstenbit Removed the duplicate conditional statement

**Review by ianstenbit — APPROVED** [review]
LGTM -- thanks!

**Pull request merged**